AI_LAB_PROJECT
Members: Muhammad Haris Hamid (24F-0699)
         Muhammad Ahmad (24F-0513)
         Waleed Bin Nasir (24F-0516)

PHASE 1
Python and Data Foundation

Step 1: Installing and importing libraries

In [4]:
!pip install pandas numpy matplotlib scikit-learn

In [5]:
#importing pandas library
import pandas as pd

Step 2: Loading Dataset

In [6]:
#loading our data set (Track D)
df=pd.read_csv("mental-heath-in-tech-2016_20161114.csv")

Step 3: Inspecting Dataset 

In [7]:
#printing the first 5 rows using head 
df.head()

,Are you self-employed?,How many employees does your company or organization have?,Is your employer primarily a tech company/organization?,Is your primary role within your company related to tech/IT?,Does your employer provide mental health benefits as part of healthcare coverage?,Do you know the options for mental health care available under your employer-provided coverage?,"Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?",Does your employer offer resources to learn more about mental health concerns and options for seeking help?,Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?,"If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:",...,"If you have a mental health issue, do you feel that it interferes with your work when being treated effectively?","If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?",What is your age?,What is your gender?,What country do you live in?,What US state or territory do you live in?,What country do you work in?,What US state or territory do you work in?,Which of the following best describes your work position?,Do you work remotely?
0,0,26-100,1.0,NaN,Not eligible for coverage / N/A,NaN,No,No,I don't know,Very easy,...,Not applicable to me,Not applicable to me,39,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Sometimes
1,0,6-25,1.0,NaN,No,Yes,Yes,Yes,Yes,Somewhat easy,...,Rarely,Sometimes,29,male,United States of America,Illinois,United States of America,Illinois,Back-end Developer|Front-end Developer,Never
2,0,6-25,1.0,NaN,No,NaN,No,No,I don't know,Neither easy nor difficult,...,Not applicable to me,Not applicable to me,38,Male,United Kingdom,NaN,United Kingdom,NaN,Back-end Developer,Always
3,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Sometimes,Sometimes,43,male,United Kingdom,NaN,United Kingdom,NaN,Supervisor/Team Lead,Sometimes
4,0,6-25,0.0,1.0,Yes,Yes,No,No,No,Neither easy nor difficult,...,Sometimes,Sometimes,43,Female,United States of America,Illinois,United States of America,Illinois,Executive Leadership|Supervisor/Team Lead|Dev ...,Sometimes


In [6]:
#printing the shape of the data set 
print(df.shape)

(1433, 63)


In [7]:
#checking missing values in the data set
df.isnull().sum()

Are you self-employed?                                                                  0
How many employees does your company or organization have?                            287
Is your employer primarily a tech company/organization?                               287
Is your primary role within your company related to tech/IT?                         1170
Does your employer provide mental health benefits as part of healthcare coverage?     287
                                                                                     ... 
What US state or territory do you live in?                                            593
What country do you work in?                                                            0
What US state or territory do you work in?                                            582
Which of the following best describes your work position?                               0
Do you work remotely?                                                                   0
Length: 63

In [9]:
#printing Unique values of target column
target_col="Have you been diagnosed with a mental health condition by a medical professional?"
print(f"Our Target column is: {target_col}")
print("Unique values in our target column are as follows :")
print(df[target_col].unique())

Our Target column is: Have you been diagnosed with a mental health condition by a medical professional?
Unique values in our target column are as follows :
<StringArray>
['Yes', 'No']
Length: 2, dtype: str


Step 4: Apply Python Fundamentals

In [10]:
#target column's different classes(response) and their count
#we define an empty dictionary to store the result 
target_col_class_count={}
for value in df[target_col]:
    #if there are any null values or blanks we convert them to string to make sure that it doesnt crash
    value=str(value)
    if value in target_col_class_count:
        target_col_class_count[value]+=1
    else:
        target_col_class_count[value]=1
print("The class Distribution in target col is as follows:")
print(target_col_class_count)        
        

The class Distribution in target col is as follows:
{'Yes': 716, 'No': 717}


In [25]:
# we store the columns of our dataset into a list
columns_list=list(df.columns)
print(columns_list)

['Are you self-employed?', 'How many employees does your company or organization have?', 'Is your employer primarily a tech company/organization?', 'Is your primary role within your company related to tech/IT?', 'Does your employer provide mental health benefits as part of healthcare coverage?', 'Do you know the options for mental health care available under your employer-provided coverage?', 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?', 'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:', 'Do you think that discussing a mental health disorder with your employer would have negative cons

In [11]:
#slicing the first 50 columns 
sample_data=df[:50]
print(sample_data)

    Are you self-employed?  \
0                        0   
1                        0   
2                        0   
3                        1   
4                        0   
5                        0   
6                        0   
7                        0   
8                        0   
9                        1   
10                       0   
11                       0   
12                       0   
13                       0   
14                       0   
15                       0   
16                       0   
17                       0   
18                       1   
19                       0   
20                       0   
21                       0   
22                       0   
23                       0   
24                       1   
25                       0   
26                       0   
27                       0   
28                       0   
29                       0   
30                       0   
31                       0   
32        

In [29]:
# function to describe the data set 
def describe_dataset():
    print("Dataset Description\n")
    print(f"Dataset Shape: {df.shape}\n")
    print(f"Dataset Column Names: {columns_list}\n")
    print(f"Dataset Class Distribution: {target_col_class_count}\n")

#calling the function
describe_dataset()
    

Dataset Description

Dataset Shape: (1433, 63)

Dataset Column Names: ['Are you self-employed?', 'How many employees does your company or organization have?', 'Is your employer primarily a tech company/organization?', 'Is your primary role within your company related to tech/IT?', 'Does your employer provide mental health benefits as part of healthcare coverage?', 'Do you know the options for mental health care available under your employer-provided coverage?', 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?', 'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:', 'Do you think that discussin

Step 5: Object-Oriented Representation

In [12]:
# we define our class DataRecord
class DataRecord:
    def __init__(self,record_id,features,label):
        self.record_id=record_id
        self.features=features
        self.label=label
    def Display(self):
        print(f"Record ID: {self.record_id}\n")
        print(f"Features: {self.features}\n")
        print(f"Label: {self.label}\n")
        print("-"*15)
#we make an empty python list to store our objects 
object_list=[]
#for loop to iterate over 5 rows
for row_index in range(5):
    #we use iloc() to grab the row with index row_index which we just defined in our for loop (eg= row0,row1,..etc) inside our row variable
    row=df.iloc[row_index]
    #we store the index of our row as record id in a new variable 
    obj_id=row_index
    #we store the label as our target column since that column is our main focus 
    obj_label=row[target_col]
    #we store rest of the columns as a separate feature in the form of a dictionary to record feature and the response of the object
    #we also drop target_col since we already stored it as our label 
    obj_features=row.drop(target_col).to_dict()
    #we pass the values into the constructor of our object 
    new_record_object=DataRecord(record_id=obj_id,label=obj_label,features=obj_features)
    #we add that item into our list of objects (object_list)
    object_list.append(new_record_object)
#we iterate over our object list and call display for each object
for obj in object_list:
    obj.Display()


Record ID: 0

Features: {'Are you self-employed?': 0, 'How many employees does your company or organization have?': '26-100', 'Is your employer primarily a tech company/organization?': 1.0, 'Is your primary role within your company related to tech/IT?': nan, 'Does your employer provide mental health benefits as part of healthcare coverage?': 'Not eligible for coverage / N/A', 'Do you know the options for mental health care available under your employer-provided coverage?': nan, 'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?': 'No', 'Does your employer offer resources to learn more about mental health concerns and options for seeking help?': 'No', 'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?': "I don't know", 'If a mental health issue prompted you to request a medical leave from work, asking for that l

Step 6: Build a Graph from Your Data

In [13]:
#we need to build a graph between 2 cols we've used are given below in col1 and col2
#the relation ship the above columns represent is how soughting help for mental health affects work if you have 
col1="Have you ever sought treatment for a mental health issue from a mental health professional?"
col2="If you have a mental health issue, do you feel that it interferes with your work when NOT being treated effectively?"
#we define an empty dictionary for our graph the dictionary's key will be the unique values of each column and the values will be 
#a list which will contain all the relationships col1 has with col2 i.e neighbors eg= col1=Yes and col2=No is one relationship 
#col1=Yes and col2=Rarely is another one so node col1 has neighbors No and Rarely 
graph={}
for index in range(len(df)):
  #we grab the current row in our dataset
    row=df.iloc[index]
 #since col1 also has 1 or 0 as vals we convert it into str to check it
    raw_val=str(row[col1]).strip()
    #if 1 we take it as a yes
    if(raw_val=='1')or (raw_val=='1.0'):
        val='yes'
    #else we take it as a no
    else:
        val='no'
    #we append these strings to differ diff types of inference and yes/no to make it sound more meaningfull
    node_A=val.strip().title()+"(Sought Treatment)"
    node_B=str(row[col2]).strip().title()+"(Interference)"
    #now we simply fill our dictionary for initial responses to setup nodes
    if (node_A not in graph):
        graph[node_A]=[]
    if (node_B not in graph):
        graph[node_B]=[]
    #we setup neighbors of each node
    if (node_B not in graph[node_A]):
         graph[node_A].append(node_B)
    if(node_A not in graph[node_B]):
        graph[node_B].append(node_A)
#we print the graph now
print("Dataset Graph\n")
for node,connected_nodes in graph.items():
    print(f"{node} is connected to -->{connected_nodes}")
#we calculate the total number of nodes and edges in the  graph
print(f"the total number of edges in the graph are {len(graph)}\n")
print(f"the total number of edges in the graph are {sum(len(connections) for connections in graph.values())//2}\n")


Dataset Graph

No(Sought Treatment) is connected to -->['Not Applicable To Me(Interference)', 'Sometimes(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)']
Not Applicable To Me(Interference) is connected to -->['No(Sought Treatment)', 'Yes(Sought Treatment)']
Yes(Sought Treatment) is connected to -->['Sometimes(Interference)', 'Not Applicable To Me(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)']
Sometimes(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Often(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Rarely(Interference) is connected to -->['Yes(Sought Treatment)', 'No(Sought Treatment)']
Never(Interference) is connected to -->['No(Sought Treatment)', 'Yes(Sought Treatment)']
the total number of edges in the graph are 7

the total number of edges in the graph are 10



                                               Phase 1 Reflection

1. How many rows and columns does your dataset have? What does each row represent?
The dataset consists of 1,433 rows and 63 columns. Each row represents a single individual, specifically a tech industry worker, and contains their specific responses to a comprehensive mental health survey.

2. Are there missing values? If yes, which columns have them and how might this affect your analysis?
Yes, there are missing values, particularly in secondary questions like the work interference columns where respondents may have skipped the question. Missing data can negatively affect the analysis because the AI might have to drop those rows entirely or guess the missing information, which can skew the accuracy of our final models.

3. What is the target variable and how many unique classes does it have?
Our target variable is the column asking, "Have you been diagnosed with a mental health condition by a medical professional?". It is a binary classification target, meaning it has exactly two unique classes: "Yes" and "No".

4. Describe the graph you built. What do the nodes and edges represent in the context of your domain?
I built an undirected bipartite graph mapping "Sought Treatment" against "Work Interference." In this domain context, the Nodes represent the unique multiple-choice answers to those questions (e.g., "Yes (Sought Treatment)" or "Often (Interference)"). The Edges represent actual human connections; an edge is only drawn between two nodes if a single employee in the dataset selected both of those specific answers.

PHASE 2: Intelligent Agent and Search Algorithms

Step 1: Define the Agent and State Space


In [14]:
print(graph.keys())

dict_keys(['No(Sought Treatment)', 'Not Applicable To Me(Interference)', 'Yes(Sought Treatment)', 'Sometimes(Interference)', 'Often(Interference)', 'Rarely(Interference)', 'Never(Interference)'])


In [18]:
#we define our agent first
class Agent:
    def __init__(self,graph,goal):
        self.graph=graph
        self.goal=goal
    def perceive(self,state):
        # Looks up the current state in the dictionary to find its neighbors
        return self.graph[state]
    def action(self,act):
        # In this simple graph, the "action" is just the name of the next node
        return act
    def goal_test(self,state):
        # Returns True if we reached the goal, otherwise False
        if(state==self.goal):
            return True
    def get_cost(self,state1,state2):
        # As per the manual, every move costs exactly 1
        return 1
test_agent = Agent(graph, "Often(Interference)")

Step 2: Uninformed Search

In [16]:
from collections import deque
#BFS
def BFS(agent, start_node):
    visited = [] 
    queue = deque([start_node]) 
    
    #The robot's memory of how it got everywhere
    came_from = {}
    came_from[start_node] = None # The start node has no parent!
    
    nodes_explored = 0
    
    while len(queue) > 0: 
        returned_node = queue.popleft()
        nodes_explored += 1 # Click the counter for your lab manual!
        
        if returned_node in visited:
            continue
        else:
            visited.append(returned_node)
            
        if agent.goal_test(returned_node): 
            print("Goal Found!")
            print(f"Total Nodes Explored: {nodes_explored}")
            
            # --- NEW: Trace the path backward! ---
            path = []
            step = returned_node
            while step is not None:
                path.append(step)
                step = came_from[step] # Look up who the parent was
                
            path.reverse() # Flip it so it goes Start -> Goal
            
            print(f"Path Length: {len(path) - 1}") # Number of steps taken
            print("Direct Path to Goal:", path)
            return path
            
        else:
            for neighbor in agent.perceive(returned_node): 
                # If we haven't seen this neighbor before, record how we got to it!
                if neighbor not in visited and neighbor not in queue:
                    came_from[neighbor] = returned_node
                queue.append(neighbor)
                
    print("goal not found")

# Test it!
BFS(test_agent, "Yes(Sought Treatment)")

Goal Found!
Total Nodes Explored: 4
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)', 'Often(Interference)']

In [21]:
#DFS
def DFS(agent,start_node):
    #visited list for tracking nodes we have alr visited 
    visited=[]
    came_from = {}
    came_from[start_node]=None 
    #a counter for number of nodes explored
    nodes_explored=0
    #a stack we made using list
    stack=[start_node]
    while len(stack)>0:
        #we pop the top of stack 
        returned_node=stack.pop()
        #incrementing the counter
        nodes_explored+=1
        #check if its in visited or not and append accordingly in visited
        if (returned_node in visited):
            continue
        else:
            visited.append(returned_node)
        if agent.goal_test(returned_node):
            print("Goal found")
            print(f"Total Nodes Explored: {nodes_explored}")
             #we construct the path using a list 
            path=[]
            step= returned_node
            while step is not None:
                path.append(step)
                step=came_from[step]
            path.reverse()
            print(f"Path Length: {len(path) - 1}") # Number of steps taken
            print("Direct Path to Goal:", path)
            return path
        else:
            for neighbor in agent.perceive(returned_node):
                #if we havent seen this neighbor push it in
                if neighbor not in visited and neighbor not in stack:
                    stack.append(neighbor)
                    came_from[neighbor]=returned_node
                    stack.append(neighbor)
    print("Goal not found")


DFS(test_agent, "Yes(Sought Treatment)")

Goal found
Total Nodes Explored: 8
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)', 'Often(Interference)']

In [26]:
#DLS
def DLS(agent, start_node, limit):
    #dictionary for tracking nodes we have alr visited and their depth
    visited = {}
    #a counter for number of nodes explored
    nodes_explored = 0
    #a stack we made using list that also stores the depth of the node
    stack = [(start_node, 0)]
    
    #dictionary to remember where we came from so we can build the path later
    came_from = {start_node: None}

    while len(stack) > 0:
        #we pop the top of stack and get both node and its depth
        returned_node, depth = stack.pop()
        #click the counter for our nodes
        nodes_explored += 1

        #check if its in visited and we found it at a shallower depth before
        if (returned_node in visited) and (visited[returned_node] <= depth):
            continue
            
        #we record the depth at which we found this node
        visited[returned_node] = depth

        if agent.goal_test(returned_node):
            print(f"Goal found at depth {depth}!")
            print(f"Total Nodes Explored: {nodes_explored}")
            
            #we make an empty list for the final path
            path = []
            #we set our starting step to the goal node we just found
            step = returned_node
            
            #we loop backwards until we hit None
            while step is not None:
                path.append(step)
                step = came_from[step]
                
            #since we built it backwards, we reverse it
            path.reverse()
            
            print(f"Path Length: {len(path) - 1}")
            print("Direct Path to Goal:", path)
            return path

        #if we havent reached the depth limit we expand
        if depth < limit:
            for neighbor in agent.perceive(returned_node):
                #if we havent seen this neighbor or we found a shorter path to it
                if (neighbor not in visited) or (visited[neighbor] > depth + 1):
                    came_from[neighbor] = returned_node
                    #push it in the stack with incremented depth
                    stack.append((neighbor, depth + 1))
                    
    print(f"Goal not found within depth limit {limit}")
    return None

DLS(test_agent,"Yes(Sought Treatment)",3)
DLS(test_agent,"Yes(Sought Treatment)",5)


Goal found at depth 3!
Total Nodes Explored: 5
Path Length: 3
Direct Path to Goal: ['Yes(Sought Treatment)', 'Never(Interference)', 'No(Sought Treatment)', 'Often(Interference)']
Goal found at depth 3!
Total Nodes Explored: 5
Path Length: 3
Direct Path to Goal: ['Yes(Sought Treatment)', 'Never(Interference)', 'No(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)',
 'Never(Interference)',
 'No(Sought Treatment)',
 'Often(Interference)']

In [25]:
#IDS
def IDS(agent, start_node, max_limit=20):
    #we loop through depth limits starting from 0 up to max_limit
    for depth_limit in range(max_limit):
        print(f"--- Testing depth limit: {depth_limit} ---")
        #we call DLS with the current depth limit
        result = DLS(agent, start_node, depth_limit)
        
        #if DLS returns a path (not None) it means we found the goal
        if result is not None:
            print(f"IDS successfully found the goal at depth limit {depth_limit}!")
            return result
            
    print("Goal not found within maximum depth limit.")
    return None
IDS(test_agent,"Yes(Sought Treatment)",20)

--- Testing depth limit: 0 ---
Goal not found within depth limit 0
--- Testing depth limit: 1 ---
Goal found at depth 1!
Total Nodes Explored: 4
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']
IDS successfully found the goal at depth limit 1!


['Yes(Sought Treatment)', 'Often(Interference)']

In [27]:
import heapq

def UCS(agent, start_node):
    #dictionary to track nodes we have visited AND the lowest cost to reach them
    visited = {}
    #a counter for number of nodes explored
    nodes_explored = 0
    
    #our priority queue using heapq. it will store tuples of (cumulative_cost, node)
    priority_queue = []
    #we push the start node with a cost of 0
    heapq.heappush(priority_queue, (0, start_node))
    
    #dictionary to remember where we came from so we can build the path later
    came_from = {start_node: None}
    
    while len(priority_queue) > 0:
        #we pop the node with the lowest cost from the priority queue
        current_cost, returned_node = heapq.heappop(priority_queue)
        #click the counter for our nodes
        nodes_explored += 1
        
        #if we already visited this node and found a cheaper or equal path before we skip it
        if returned_node in visited and visited[returned_node] <= current_cost:
            continue
            
        #we record the lowest cost to reach this node
        visited[returned_node] = current_cost
        
        if agent.goal_test(returned_node):
            print("Goal found!")
            print(f"Total Nodes Explored: {nodes_explored}")
            print(f"Total Path Cost: {current_cost}")
            
            #we make an empty list for the final path
            path = []
            #we set our starting step to the goal node we just found
            step = returned_node
            
            #we loop backwards until we hit None
            while step is not None:
                path.append(step)
                step = came_from[step]
                
            #since we built it backwards, we reverse it
            path.reverse()
            
            print(f"Path Length: {len(path) - 1}")
            print("Direct Path to Goal:", path)
            return path
            
        else:
            for neighbor in agent.perceive(returned_node):
                #we calculate the cost to reach this neighbor
                step_cost = agent.get_cost(returned_node, neighbor)
                new_cost = current_cost + step_cost
                
                #if we havent seen this neighbor or we found a cheaper path to it
                if neighbor not in visited or new_cost < visited.get(neighbor, float('inf')):
                    #record that we got to this neighbor from the returned_node
                    came_from[neighbor] = returned_node
                    #push it into the priority queue with its new total cost
                    heapq.heappush(priority_queue, (new_cost, neighbor))
                    
    print("Goal not found")
    return None

UCS(test_agent, "Yes(Sought Treatment)")

Goal found!
Total Nodes Explored: 4
Total Path Cost: 1
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)', 'Often(Interference)']

In [28]:
#a simple heuristic function to estimate cost to goal
def heuristic(state, goal):
    #if we are already at the goal the cost is 0
    if state == goal:
        return 0
        
    #we treat the words inside the node string as features
    #we replace brackets with spaces and split into words
    state_features = set(state.replace("(", " ").replace(")", " ").split())
    goal_features = set(goal.replace("(", " ").replace(")", " ").split())
    
    #we count how many exact words match between current state and goal
    matches = len(state_features.intersection(goal_features))
    
    #we invert it so more matches give a lower score (closer to goal)
    return len(goal_features) - matches

In [31]:
import heapq

#Best First Search
def BestFirstSearch(agent, start_node):
    #visited list for tracking nodes we have alr visited
    visited = []
    #a counter for number of nodes explored
    nodes_explored = 0
    
    #our priority queue stores tuples of heuristic value and node
    priority_queue = []
    #we calculate heuristic for start node and push it
    start_h = heuristic(start_node, agent.goal)
    heapq.heappush(priority_queue, (start_h, start_node))
    
    #dictionary to remember where we came from so we can build the path later
    came_from = {start_node: None}
    
    while len(priority_queue) > 0:
        #we pop the node with lowest heuristic value
        current_h, returned_node = heapq.heappop(priority_queue)
        #click the counter for our nodes
        nodes_explored += 1
        
        #check if its in visited or not and append accordingly
        if returned_node in visited:
            continue
        else:
            visited.append(returned_node)
            
        if agent.goal_test(returned_node):
            print("Goal found!")
            print(f"Total Nodes Explored: {nodes_explored}")
            
            #we make an empty list for the final path
            path = []
            step = returned_node
            
            #we loop backwards until we hit None
            while step is not None:
                path.append(step)
                step = came_from[step]
                
            #since we built it backwards we reverse it
            path.reverse()
            
            print(f"Path Length: {len(path) - 1}")
            print("Direct Path to Goal:", path)
            return path
            
        else:
            for neighbor in agent.perceive(returned_node):
                #if we havent seen this neighbor push it in
                if neighbor not in visited:
                    came_from[neighbor] = returned_node
                    #calculate heuristic for neighbor and push it
                    neighbor_h = heuristic(neighbor, agent.goal)
                    heapq.heappush(priority_queue, (neighbor_h, neighbor))
                    
    print("Goal not found")
    return None
BestFirstSearch(test_agent, "Yes(Sought Treatment)")

Goal found!
Total Nodes Explored: 2
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)', 'Often(Interference)']

In [32]:
#A* Search
def AStarSearch(agent, start_node):
    #we track the lowest actual path cost to reach each node
    g_costs = {start_node: 0}
    #a counter for number of nodes explored
    nodes_explored = 0
    
    #priority queue stores f_cost and node where f = g + h
    priority_queue = []
    #initial f_cost is just 0 + heuristic
    start_h = heuristic(start_node, agent.goal)
    heapq.heappush(priority_queue, (start_h, start_node))
    
    #dictionary to remember where we came from
    came_from = {start_node: None}
    
    while len(priority_queue) > 0:
        #pop the node with the lowest f_cost
        current_f, returned_node = heapq.heappop(priority_queue)
        #click the counter for our nodes
        nodes_explored += 1
        
        #check goal test after popping for shortest path guarantee
        if agent.goal_test(returned_node):
            print("Goal found!")
            print(f"Total Nodes Explored: {nodes_explored}")
            print(f"Total Path Cost: {g_costs[returned_node]}")
            
            #we make an empty list for the final path
            path = []
            step = returned_node
            
            #we loop backwards until we hit None
            while step is not None:
                path.append(step)
                step = came_from[step]
                
            #since we built it backwards we reverse it
            path.reverse()
            
            print(f"Path Length: {len(path) - 1}")
            print("Direct Path to Goal:", path)
            return path
            
        for neighbor in agent.perceive(returned_node):
            #calculate the new actual cost to reach this neighbor
            step_cost = agent.get_cost(returned_node, neighbor)
            new_g = g_costs[returned_node] + step_cost
            
            #if we found a cheaper path or havent seen it yet
            if neighbor not in g_costs or new_g < g_costs[neighbor]:
                #record the new cheaper g_cost
                g_costs[neighbor] = new_g
                
                #calculate f = g + h
                f_cost = new_g + heuristic(neighbor, agent.goal)
                
                #record where we came from and push to queue
                came_from[neighbor] = returned_node
                heapq.heappush(priority_queue, (f_cost, neighbor))
                
    print("Goal not found")
    return None
AStarSearch(test_agent, "Yes(Sought Treatment)")

Goal found!
Total Nodes Explored: 2
Total Path Cost: 1
Path Length: 1
Direct Path to Goal: ['Yes(Sought Treatment)', 'Often(Interference)']


['Yes(Sought Treatment)', 'Often(Interference)']

In [33]:
import random

#Hill Climbing
def HillClimbing(agent, start_node):
    #we start at the given node
    current_node = start_node
    #a counter for number of steps taken
    steps = 0
    
    while True:
        #click the counter for our steps
        steps += 1
        
        #if we reached the goal we return it
        if agent.goal_test(current_node):
            #returning True means we hit the global optimum (the goal)
            return True, current_node
            
        #we get all neighbors of current node
        neighbors = agent.perceive(current_node)
        
        #we find the neighbor with the best (lowest) heuristic score
        #we initialize with a very high number
        best_h = float('inf')
        best_neighbor = None
        
        for neighbor in neighbors:
            h = heuristic(neighbor, agent.goal)
            if h < best_h:
                best_h = h
                best_neighbor = neighbor
                
        #we check the score of where we are currently standing
        current_h = heuristic(current_node, agent.goal)
        
        #if the best neighbor is worse or equal to where we are we have hit a peak
        if best_h >= current_h:
            #returning False means we got stuck in a local optimum
            return False, current_node
            
        #otherwise we move to the best neighbor
        current_node = best_neighbor

#--- Code to test Hill Climbing 10 times as per manual ---
print("--- Hill Climbing 10 Runs ---")
#we grab all unique nodes from the graph to use as random start points
all_nodes = list(test_agent.graph.keys())

success_count = 0
for i in range(10):
    random_start = random.choice(all_nodes)
    reached_goal, final_node = HillClimbing(test_agent, random_start)
    
    if reached_goal:
        success_count += 1
        print(f"Run {i+1}: SUCCESS! Reached Goal from {random_start}")
    else:
        print(f"Run {i+1}: STUCK at local optimum {final_node} from {random_start}")

print(f"Global Optimum reached {success_count}/10 times.")

--- Hill Climbing 10 Runs ---
Run 1: SUCCESS! Reached Goal from Yes(Sought Treatment)
Run 2: STUCK at local optimum Not Applicable To Me(Interference) from Not Applicable To Me(Interference)
Run 3: STUCK at local optimum Sometimes(Interference) from Sometimes(Interference)
Run 4: STUCK at local optimum Never(Interference) from Never(Interference)
Run 5: STUCK at local optimum Never(Interference) from Never(Interference)
Run 6: STUCK at local optimum Sometimes(Interference) from Sometimes(Interference)
Run 7: STUCK at local optimum Sometimes(Interference) from Sometimes(Interference)
Run 8: SUCCESS! Reached Goal from Often(Interference)
Run 9: STUCK at local optimum Sometimes(Interference) from Sometimes(Interference)
Run 10: STUCK at local optimum Never(Interference) from Never(Interference)
Global Optimum reached 2/10 times.


In [37]:
import math

#Simulated Annealing
def SimulatedAnnealing(agent, start_node):
    current_node = start_node
    #temperature starts at 100 as per manual
    temperature = 100.0
    #cooling rate as per manual
    cooling_rate = 0.95
    steps = 0
    
    #we run until the temperature gets very close to 0
    while temperature > 0.1:
        steps += 1
        
        if agent.goal_test(current_node):
            print(f"Goal found in {steps} steps!")
            return current_node
            
        neighbors = agent.perceive(current_node)
        if not neighbors:
            return current_node
            
        #unlike hill climbing we pick a completely random neighbor
        next_node = random.choice(neighbors)
        
        current_h = heuristic(current_node, agent.goal)
        next_h = heuristic(next_node, agent.goal)
        
        #we calculate the change in value 
        #a lower heuristic is better so we subtract next from current
        delta_e = current_h - next_h
        
        #if delta_e is positive the next node is better so we definitely move
        if delta_e > 0:
            current_node = next_node
        else:
            #if it is worse we might still move there based on a probability formula
            probability = math.exp(delta_e / temperature)
            #we pick a random number between 0 and 1
            if random.random() < probability:
                current_node = next_node
                
        #we cool down the temperature by multiplying by 0.95
        temperature = temperature * cooling_rate
        
    print(f"Ended at {current_node} after temperature cooled down")
    return current_node
SimulatedAnnealing(test_agent, "Yes(Sought Treatment)")

Goal found in 6 steps!


'Often(Interference)'

In [36]:
#Local Beam Search
def LocalBeamSearch(agent, k):
    #we grab all nodes to pick k random starting points
    all_nodes = list(agent.graph.keys())
    
    #we start with a list of k random states
    #we use random.sample to make sure we don't pick the exact same node twice
    current_states = random.sample(all_nodes, k)
    steps = 0
    
    print(f"Starting Beam Search with k={k} at states: {current_states}")
    
    while True:
        steps += 1
        
        #check if any of our current states is the goal
        for state in current_states:
            if agent.goal_test(state):
                print(f"Goal found in {steps} steps!")
                return state
                
        #we gather all neighbors for all k states into one big pool
        all_neighbors = []
        for state in current_states:
            neighbors = agent.perceive(state)
            all_neighbors.extend(neighbors)
            
        #we remove duplicates to avoid checking the same node multiple times
        all_neighbors = list(set(all_neighbors))
        
        #we sort all generated neighbors based on their heuristic value
        #we use a lambda function to tell it to sort by the heuristic score
        all_neighbors.sort(key=lambda node: heuristic(node, agent.goal))
        
        #we keep only the top k neighbors for the next round
        current_states = all_neighbors[:k]
        
        #safety break if we run out of neighbors
        if not current_states:
            print("No more states to explore")
            return None
            
#--- Code to test Beam Search with k=3 and k=5 as per manual ---
print("\n--- Beam Search (k=3) ---")
LocalBeamSearch(test_agent, k=3)

print("\n--- Beam Search (k=5) ---")
LocalBeamSearch(test_agent, k=5)


--- Beam Search (k=3) ---
Starting Beam Search with k=3 at states: ['Sometimes(Interference)', 'Often(Interference)', 'Not Applicable To Me(Interference)']
Goal found in 1 steps!

--- Beam Search (k=5) ---
Starting Beam Search with k=5 at states: ['Yes(Sought Treatment)', 'Rarely(Interference)', 'Never(Interference)', 'No(Sought Treatment)', 'Not Applicable To Me(Interference)']
Goal found in 2 steps!


'Often(Interference)'

In [38]:
#Minimax Search
def Minimax(agent, state, depth, is_max_player, nodes_evaluated):
    #click the counter for our nodes evaluated
    nodes_evaluated[0] += 1
    
    #if we reached depth limit or found the goal we stop and evaluate
    if depth == 0 or agent.goal_test(state):
        #our heuristic is distance to goal (lower is better)
        #we subtract from 10 to make it a score where higher is better for MAX
        score = 10 - heuristic(state, agent.goal)
        return score, state
        
    neighbors = agent.perceive(state)
    #if there are no neighbors we just evaluate where we are
    if not neighbors:
        return 10 - heuristic(state, agent.goal), state
        
    best_move = None
    
    #if it is our agents turn we want to maximize the score
    if is_max_player:
        best_score = -float('inf')
        for neighbor in neighbors:
            #we call minimax for the opponents turn (is_max_player becomes False)
            eval_score, _ = Minimax(agent, neighbor, depth - 1, False, nodes_evaluated)
            
            #if this move gets us a better score we update it
            if eval_score > best_score:
                best_score = eval_score
                best_move = neighbor
                
        #print the chosen move at each level as per manual requirement
        print(f"Depth {depth} (MAX) chose move: {best_move} with score: {best_score}")
        return best_score, best_move
        
    #if it is the opponents turn they want to minimize our score
    else:
        best_score = float('inf')
        for neighbor in neighbors:
            #we call minimax for our turn (is_max_player becomes True)
            eval_score, _ = Minimax(agent, neighbor, depth - 1, True, nodes_evaluated)
            
            #if this move gets them a lower score they update it
            if eval_score < best_score:
                best_score = eval_score
                best_move = neighbor
                
        #print the chosen move at each level
        print(f"Depth {depth} (MIN) chose move: {best_move} with score: {best_score}")
        return best_score, best_move

In [ ]:
#Alpha Beta Pruning
def AlphaBeta(agent, state, depth, alpha, beta, is_max_player, nodes_evaluated):
    #click the counter for our nodes evaluated
    nodes_evaluated[0] += 1
    
    #base case is the same as minimax
    if depth == 0 or agent.goal_test(state):
        score = 10 - heuristic(state, agent.goal)
        return score, state
        
    neighbors = agent.perceive(state)
    if not neighbors:
        return 10 - heuristic(state, agent.goal), state
        
    best_move = None
    
    if is_max_player:
        best_score = -float('inf')
        for neighbor in neighbors:
            eval_score, _ = AlphaBeta(agent, neighbor, depth - 1, alpha, beta, False, nodes_evaluated)
            
            if eval_score > best_score:
                best_score = eval_score
                best_move = neighbor
                
            #update alpha to the highest value we can guarantee so far
            alpha = max(alpha, best_score)
            
            #if beta is less than or equal to alpha the opponent will block this branch
            if beta <= alpha:
                #we prune the tree and break out of the loop early
                break
                
        return best_score, best_move
        
    else:
        best_score = float('inf')
        for neighbor in neighbors:
            eval_score, _ = AlphaBeta(agent, neighbor, depth - 1, alpha, beta, True, nodes_evaluated)
            
            if eval_score < best_score:
                best_score = eval_score
                best_move = neighbor
                
            #update beta to the lowest value the opponent can guarantee
            beta = min(beta, best_score)
            
            #if beta is less than or equal to alpha we would never pick this branch
            if beta <= alpha:
                #we prune the tree
                break
                
        return best_score, best_move